### PR TITLE
Make client/request an explicit parameter

### PR DESCRIPTION
**Pull request by frenchy64**
Related https://github.com/threatgrid/iroh/issues/4440

Also made a few enhancements like deleting clj-http-fake and medley and uniquifying all indices and aliases.

Reviewers: recommend hiding whitespace changes.

### CODE REVIEW

File: src/ductile/document.clj [hunk 1]
```diff
@@ -95,16 +94,18 @@
...
-                    (when (= version 5) doc-type)
-                    id)
-       (client/get (conn/make-http-opts cm opts [:_source]))
+   (-> (conn/make-http-opts cm opts [:_source])
+       (assoc :method :get
+              :url (get-doc-uri uri
+                                index-name
+                                (when (= version 5) doc-type)
+                                id))
+       request-fn
```
Comment (ereteog) [inline src/ductile/document.clj:108]: it's cleaner, thxs!

File: src/ductile/schemas.clj [hunk 2]
```diff
@@ -5,21 +5,28 @@
   (:import [org.apache.http.impl.conn PoolingClientConnectionManager
             PoolingHttpClientConnectionManager]))
 
+(s/defschema RequestFn
+  "A function implementing the 1-argument
+  arity of clj-http.client/request."
+  (s/=> s/Any (s/named s/Any 'req)))
```
Comment (ereteog) [inline src/ductile/schemas.clj:11]: we could perhaps have an open schema with the minimal set of keys that we do use
```clojure
(st/open-schema
  {:url s/Str
   :method (s/enum :get :post :put :delete :head)
   :body (s/maybe {s/Any s/Any})
   :form-params  (s/maybe {s/Any s/Any})
   :query-params  (s/maybe {s/Any s/Any})
   :connection-manager (s/either PoolingClientConnectionManager PoolingHttpClientConnectionManager)}
```
in addtion the defaut opts being:
```clojure
(def default-opts
  {:as :json
   :content-type :json
   :throw-exceptions false})
```
The return schema  could be  (s/maybe {s/Any s/Any}) (I propose this schema for body, form-params and query-params for the same reason)

File: test/ductile/document_test.clj [hunk 3]
```diff
@@ -149,440 +150,448 @@
...
-                                          q-ids-2
-                                          {:wait_for_completion false
-                                           :refresh "true"})))
-         "delete-by-query with wait-for-completion? set to false should directly return an answer before deletion with a task id"))))
+  (let [indexname (str "test_index" (UUID/randomUUID))
+        indexname1 (str indexname "-1")
+        indexname2 (str indexname "-2")]
+    (for-each-es-version
+     "delete-by-query should trigger a proper _delete_by_query request"
+     #(es-index/delete! conn indexname)
```
Comment (ereteog) [inline test/ductile/document_test.clj:478]: We should add a wildward to ensure that "indexname-1" and "indexname-2" are actually cleaned if the `delete-by-query` failed in the tests (I relied on it, but it's better to add the wildcard too):
```suggestion
     #(es-index/delete! conn (str indexname "*")
```

File: test/ductile/document_test.clj [hunk 4]
```diff
@@ -149,440 +150,448 @@
...
-                            "`refresh` is the only accepted query parameter")))))
+                             (reset! query-params nil))
+        conn (es-conn/connect {:host "localhost"
+                               :port 9200
+                               :request-fn
+                               (-> (fn [req]
+                                     (reset! query-params (keywordize-keys (form-decode (:query-string req))))
+                                     {:status 200
+                                      :headers {:content-type "application/clojure"}})
+                                   client/wrap-query-params)})]
```
Comment (ereteog) [inline test/ductile/document_test.clj:536]: nice!

**Review by ereteog — APPROVED** [review]
Very Nice!
Could you update the README and add a recommendation for faking the request to ES?

File: test/ductile/document_test.clj (hunk 3 above)
Comment (frenchy64) [reply test/ductile/document_test.clj:478]: Makes sense. I was wondering what kind of magic made those indices get deleted.

File: src/ductile/schemas.clj (hunk 2 above)
Comment (frenchy64) [reply src/ductile/schemas.clj:11]: For some reason this makes me nervous so I'll give it some time to sink in.

**Pull request merged**